src/de.rs

```diff
@@ -305,59 +275,69 @@ impl<'de, 'a> de::Deserializer<'de> for &'a mut Deserializer<'de> {
     fn deserialize_option<V>(self, visitor: V) -> Result<V::Value>
         where V: Visitor<'de>
     {
-        if self.input.starts_with("null") {
-            self.input = &self.input["null".len()..];
-            visitor.visit_none()
-        } else {
-            visitor.visit_some(self)
+        match seq(b"null").discard().parse(&mut self.input) {
+            Ok(_) => visitor.visit_none(),
+            Err(_) => visitor.visit_some(self),
         }
     }
 
     // In Serde, unit means an anonymous value containing no data.
     fn deserialize_unit<V>(self, visitor: V) -> Result<V::Value>
         where V: Visitor<'de>
     {
-        if self.consume("()") {
-            visitor.visit_unit()
-        } else {
-            Err(Error::ExpectetUnit)
+        match self.consume("()") {
+            Ok(_) => visitor.visit_unit(),
+            Err(_) => Err(Error::ExpectetUnit),
         }
     }
 
     fn deserialize_unit_struct<V>(
         self,
-        _name: &'static str,
+        name: &'static str,
         visitor: V
     ) -> Result<V::Value>
         where V: Visitor<'de>
     {
-        // TODO: Should we also allow `MyUnitValue`?
-
-        self.deserialize_unit(visitor)
+        match self.consume(name) {
+            Ok(_) => visitor.visit_unit(),
+            Err(_) => Err(Error::ExpectedStructName),
```

Comment:
I see `()` as already being overloaded. It can be used to represent a unit, but also - an empty non-unit struct (defined as `MyStruct{}`), since we can omit struct names, and their parenthesis are `(`/`)`.